classes/invitation/invitations/userRoleAssignment/UserRoleAssignmentInvite.php

```diff
@@ -0,0 +1,411 @@
+<?php
+
+/**
+ * @file classes/invitation/invitations/UserRoleAssignmentInvite.php
+ *
+ * Copyright (c) 2024 Simon Fraser University
+ * Copyright (c) 2024 John Willinsky
+ * Distributed under the GNU GPL v3. For full terms see the file docs/COPYING.
+ *
+ * @class UserRoleAssignmentInvite
+ *
+ * @brief Assign Roles to User invitation
+ */
+
+namespace PKP\invitation\invitations\userRoleAssignment;
+
+use APP\core\Application;
+use APP\facades\Repo;
+use Exception;
+use Illuminate\Mail\Mailable;
+use PKP\identity\Identity;
+use PKP\invitation\core\contracts\IApiHandleable;
+use PKP\invitation\core\CreateInvitationController;
+use PKP\invitation\core\enums\InvitationStatus;
+use PKP\invitation\core\Invitation;
+use PKP\invitation\core\InvitationActionRedirectController;
+use PKP\invitation\core\ReceiveInvitationController;
+use PKP\invitation\core\traits\HasMailable;
+use PKP\invitation\core\traits\ShouldValidate;
+use PKP\invitation\invitations\userRoleAssignment\handlers\api\UserRoleAssignmentCreateController;
+use PKP\invitation\invitations\userRoleAssignment\handlers\api\UserRoleAssignmentReceiveController;
+use PKP\invitation\invitations\userRoleAssignment\handlers\UserRoleAssignmentInviteRedirectController;
+use PKP\invitation\invitations\userRoleAssignment\payload\UserGroupPayload;
+use PKP\invitation\models\InvitationModel;
+use PKP\mail\mailables\UserRoleAssignmentInvitationNotify;
+use PKP\userGroup\relationships\UserUserGroup;
+
+class UserRoleAssignmentInvite extends Invitation implements IApiHandleable
+{
+    use HasMailable;
+    use ShouldValidate;
+
+    public const INVITATION_TYPE = 'userRoleAssignment';
+
+    protected array $notAccessibleAfterInvite = [
+        'userGroupsToAdd',
+        'userGroupsToRemove',
+    ];
+
+    protected array $notAccessibleBeforeInvite = [
+        'orcid',
+        'username',
+        'password'
+    ];
+
+    public ?string $orcid = null;
+    public ?string $givenName = null;
+    public ?string $familyName = null;
+    public ?string $affiliation = null;
+    public ?string $country = null;
+
+    public ?string $username = null;
+    public ?string $password = null;
+
+    public ?string $emailSubject = null;
+    public ?string $emailBody = null;
+    public ?bool $existingUser = null;
+
+    public ?array $userGroupsToAdd = null;
```

Comment:
It looks like `$userGroupsToAdd` / `$userGroupsToRemove` are sometimes arrays of `UserGroup` objects, and other times arrays of `UserGroupPayload` objects. They should be one consistent type.